README.md
# asdig
simple unix util for retrieving AS (autonomous system) match from IP(v6) published
by services like Rezopole GoASMap (http://github.com/rezopole/goasmap) or cymru.

## Usage
how to retrieve the AS that publish a prefix for 134.214.100.6 ?

```bash
$ asdig 134.214.100.6
2060 | 134.214.0.0/16 | EU | ripe | 1993-09-01 | FR-RENATER
```

AS-2060 (RENATER) does via the 134.214.0.0/16 subnet route.

who publish some IPv6 route for www.nasa.gov ?

```bash
$ host www.nasa.gov
www.nasa.gov is an alias for www.nasawestprime.com.
www.nasawestprime.com is an alias for iznasa.hs.llnwd.net.
iznasa.hs.llnwd.net has address 95.140.230.145
iznasa.hs.llnwd.net has IPv6 address 2a02:3d0:623:a000::8008

$ asdig 2a02:3d0:623:a000::8008
22822 | 2a02:3d0::/32 | US | arin | 2001-11-28 | LLNW
```

AS-22822 (LLWN) does via the 2a02:3d0::/32 subnet announce.

who is AS-199422 ?

```bash
$ asdig as199422
199422 | FR | ripe | 2013-01-24 | REZOPOLE
```

AS-199422 is registered from RIPE to some french organisation (since 2013-01-24) that named it "REZOPOLE".

## Compilation / Installation
the following should bring you to a traditional build tree :
```bash
   libtoolize --force --copy && aclocal && autoheader && automake --add-missing --copy && autoconf
```
then the traditional build sequence :
```bash
    ./configure
    make
    make install
```

## Prerequisites
an autoconf / automake build-chain, in order to get a compilable tree.

a C++ compiler and the resolv library with it's headers.

## Authors
The Rezopole team (https://www.rezopole.net/en/about-rezopole/team)



debian-proto/changelog
asdig (0.0.97-1) RELEASE-TARGET; urgency=medium

  * Minor corrections

 -- Jean-Daniel Pauget <sysadm@rezopole.net>  Sun, 17 May 2020 18:00:00 +0200
asdig (0.0.96-1) RELEASE-TARGET; urgency=medium

  * Initial release

 -- Jean-Daniel Pauget <sysadm@rezopole.net>  Thu, 14 May 2020 20:39:37 +0200


ethertype.h

// {
//
//  asdig, retreives Autonomous System (AS) infos from repositories via DNS rev-queries 
//  Copyright (C) 2016 Jean-Daniel Pauget <jdpauget@rezopole.net>
//  
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//  
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//  
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// }




#ifndef RZPNETETHERTYPE
#define RZPNETETHERTYPE

#include <iostream>
#include <string>

#if defined(__OpenBSD__)
#include <net/ethertypes.h>
#else
#include <net/ethernet.h>
#endif

// --------- Ethertype -------------------------------------------------------------------------------------------------------------------------

namespace rzpnet {

    using namespace std;

typedef enum {
    TETHER_IPV4,
    TETHER_IPV6,
    TETHER_ARP,
    TETHER_8023,
    TETHER_8021Q,
    TETHER_MOPRC,
    TETHER_LLC,
    TETHER_AOE,
    TETHER_UNKNOWN,
    TETHER_OTHER
} TEthertype;

class Ethertype {
  public:
    TEthertype ethertype;
    Ethertype () : ethertype (TETHER_UNKNOWN) {};
    Ethertype (const u_short *p) {
	switch (ntohs(*p)) {
	    case ETHERTYPE_IP:		ethertype = TETHER_IPV4;	    break;
	    case ETHERTYPE_ARP:		ethertype = TETHER_ARP;		    break;
	    case ETHERTYPE_VLAN:	ethertype = TETHER_8021Q;	    break;
	    case ETHERTYPE_IPV6:	ethertype = TETHER_IPV6;	    break;
	    default:			ethertype = TETHER_OTHER;	    break;
	}
    }
    Ethertype (const string &s) {
//cerr << "===================" << s.substr (0,20) << endl;
	     if (s.find ("ethertype IPv4") == 0)		ethertype = TETHER_IPV4;
	else if (s.find ("ethertype IPv6") == 0)		ethertype = TETHER_IPV6;
	else if (s.find ("ethertype ARP") == 0)			ethertype = TETHER_ARP;
	else if (s.find ("802.3") == 0)				ethertype = TETHER_8023;
	else if (s.find ("ethertype 802.1Q") == 0)		ethertype = TETHER_8021Q;
	else if (s.find ("LLC") == 0)				ethertype = TETHER_LLC;
	else if (s.find ("ethertype MOP RC") == 0)		ethertype = TETHER_MOPRC;
	else if (s.find ("ethertype AoE") == 0)			ethertype = TETHER_AOE;
	else if (s.find ("ethertype Unknown ") == 0)		ethertype = TETHER_UNKNOWN;
	else							ethertype = TETHER_OTHER;
    }
    Ethertype (const Ethertype &o) : ethertype(o.ethertype) {}
    bool operator< (const Ethertype &a) const {
	return ethertype < a.ethertype;
    }
};
ostream &operator<< (ostream &out, const Ethertype &p) {
    switch (p.ethertype) {
	case TETHER_IPV4:    return out << "IPv4";
	case TETHER_IPV6:    return out << "IPv6";
	case TETHER_ARP:     return out << "ARP";
	case TETHER_8023:    return out << "803.3";
	case TETHER_MOPRC:   return out << "MOP RC";
	case TETHER_AOE:     return out << "AoE";
	case TETHER_8021Q:   return out << "802.1q";
	case TETHER_UNKNOWN: return out << "Unknown";
	case TETHER_OTHER:   return out << "Other";
	default:             return out << "Other";
    }
}


} // namespace rzp_net

#endif //  RZPNETETHERTYPE


level3addr.h

// {
//
//  asdig, retreives Autonomous System (AS) infos from repositories via DNS rev-queries 
//  Copyright (C) 2016 Jean-Daniel Pauget <jdpauget@rezopole.net>
//  
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//  
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//  
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// }



#ifndef RZPNETLEVEL3ADDR
#define RZPNETLEVEL3ADDR

#include <ostream>
#include <sstream>

#include <arpa/inet.h>	// inet_pton
#include <stdlib.h>	// atoi

#include "ethertype.h"

namespace rzpnet {

    using namespace std;
    using namespace rzpnet;

// --------- Level3Addr ------------------------------------------------------------------------------------------------------------------------

class Level3Addr;
ostream &operator<< (ostream &out, const Level3Addr &a);
Level3Addr l3mask (int nb);

class Level3Addr {
  private:
    static Level3Addr mask[128];
    static bool mask_not_initialized;

  public:
    TEthertype t;
    unsigned char b[16];

    bool valid (void) const {
	return ((t == TETHER_IPV4) || (t == TETHER_IPV6));
    }

    Level3Addr (void)  : t(TETHER_UNKNOWN) {
	size_t i;
	for (i=0 ; i<16 ; i++) b[i] = 0;
    }

    void applymask (Level3Addr const &mask) {
	size_t i;
	for (i=0 ; i<16 ; i++) b[i] &= mask.b[i];
    }

    inline void init_mask (void) {
	if (mask_not_initialized) {
	    size_t i;
	    for (i=0 ; i<128 ; i++) {
		mask[i] = l3mask(i);
	    }
	    mask_not_initialized = false;
	}
    }

    void applymask (int masklen) {
	init_mask();
	if (masklen < 0) masklen = 0;
	if (masklen >=128) return;
	applymask (mask[masklen]);
    }

    Level3Addr (TEthertype proposed_type, const u_char* p) : t(TETHER_UNKNOWN) {
	size_t i;
	switch (proposed_type) {
	  case TETHER_IPV4:
	    for (i=0 ; i<4 ; i++) b[i] = *p++;
	    for ( ; i<16 ; i++) b[i] = 0;
	    t = TETHER_IPV4;
	    break;
	  case TETHER_IPV6:
	    for (i=0 ; i<16 ; i++) b[i] = *p++;
	    t = TETHER_IPV6;
	    break;
	  default:
	    for (i=0 ; i<16 ; i++) b[i] = 0;
	    t = TETHER_UNKNOWN;
	}
    }

    Level3Addr (TEthertype proposed_type, string const &s) : t(TETHER_UNKNOWN) {
	size_t p, i;
	switch (proposed_type) {
	  case TETHER_IPV4:
	    b[0] = atoi (s.c_str());
	    p = s.find('.');
	    if (p!=string::npos) {
		b[1] = atoi (s.substr(p+1).c_str());
		p = s.find('.', p+1);
		if (p!=string::npos) {
		    b[2] = atoi (s.substr(p+1).c_str());
		    p = s.find('.', p+1);
		    if (p!=string::npos) {
			b[3] = atoi (s.substr(p+1).c_str());
			t = TETHER_IPV4;
		    }
		}
	    }
	    for (i=4 ; i<16 ; i++) b[i] = 0;	// we pad with zeros for comparisons
	    break;

	  case TETHER_IPV6:
	    p = s.find_first_not_of ("0123456789abcdefABCDEF:");
	    if (p == string::npos) {
		if (inet_pton(AF_INET6, s.c_str(), &(b[0])) == 1)
		    t = TETHER_IPV6;
	    } else {
		if (inet_pton(AF_INET6, s.substr(0,p).c_str(), &(b[0])) == 1)
		    t = TETHER_IPV6;
	    }
	    break;

	  default:
	    for (i=0 ; i<16 ; i++) b[i] = 0;
	    t = TETHER_UNKNOWN;
	}
    }

    Level3Addr (Level3Addr const &r) : t(r.t) {
	size_t i;
	for (i=0 ; i<16 ; i++) b[i] = r.b[i];
    }
    bool operator< (const Level3Addr &a) const {
	if (t < a.t) {
	    return true;
	} else if (a.t < t) {
	    return false;
	} else {	// we have equal types
	    size_t i;
	    for (i=0 ; i<16 ; i++) {
		if (b[i] < a.b[i]) {
		    return true;
		} else if (a.b[i] < b[i]) {
		    return false;
		}
	    }
	}
	return false;	// we should get there only if a == *this
    }

    bool operator== (const Level3Addr &a) const {
	if (t != a.t) {
	    return false;
	}
	// we have equal types
	size_t i;
	for (i=0 ; i<16 ; i++) {
	    if (b[i] != a.b[i]) {
		return false;
	    }
	}
	return true;	// we should get there only if a == *this
    }

    int rev_arpa_radix (ostream & cout) {
	int i;
	const char *h = "0123456789abcdef";
	switch (t) {
	    case TETHER_IPV4:
		for (i=3 ; i>=0 ; i--) {
		    cout << (int) b[i];
		    if (i>0)
		    cout << '.';
		}
		break;
	    case TETHER_IPV6:
		for (i=15 ; i>=0 ; i--) {
		    cout << h[ b[i] & 0xf ] << '.' << h[ (b[i] & 0xf0)>>4 ];
		    if (i>0)
		    cout << '.';
		}
		break;
	    default:
		return -1;
	}
	return 0;
    }
};

bool Level3Addr::mask_not_initialized = true;
Level3Addr Level3Addr::mask[128];

ostream &operator<< (ostream &out, const Level3Addr &a) {
    stringstream s;
    switch (a.t) {
      case TETHER_IPV4:
	s << (unsigned int)a.b[0] << '.' << (unsigned int)a.b[1] << '.' << (unsigned int)a.b[2] << '.' << (unsigned int)a.b[3];
	return out << s.str();
	break;
      case TETHER_IPV6:
	char str[INET6_ADDRSTRLEN];
	inet_ntop(AF_INET6, &(a.b[0]), str, INET6_ADDRSTRLEN);
	return out << str;
	break;
      default:
	break;
    }
    return out << "invalidL3addr";
}

Level3Addr l3mask (int nb) {
    Level3Addr mask;
    mask.t = TETHER_IPV6;

    int bit = 0;

    for (bit=0 ; bit<128 ; bit++) {
	size_t i = bit / 8;
	size_t n = 7-(bit % 8);

	if (bit < nb) {	// we set one bit to 1
	    mask.b[i] |= (1 << n);
	} else {	// we set one bit to 0
	    mask.b[i] &= ~(1 << n);
	}
    }
    return mask;
}


} // namespace rzp_net

#endif // RZPNETLEVEL3ADDR



debian-proto.sh
#!/bin/bash

    OURPACKAGENAME=$( grep '^AC_INIT' ./configure.ac | cut -d'[' -f2 | cut -d']' -f1 )
    OURPACKAGEVERSION=$( grep '^AC_INIT' ./configure.ac | cut -d'[' -f3 | cut -d']' -f1 )
    RELEASE=$( lsb_release -c -s )

    OURBUILDDIR=$( mktemp -d /tmp/"${OURPACKAGENAME}"_build_XXXXX  )
    FINALDEST=$( pwd )

    DEBHELPERVERSION=$( dpkg -s debhelper | grep '^Version:' | sed 's/^.*sion: //' | cut -d. -f1 )

    if [ ! -d "${OURBUILDDIR}" ]
    then
	echo 'mktemp -d /tmp/'"${OURPACKAGENAME}"'_build_XXXXX failed !!'
	exit 1
    else
	echo "will build  ${OURPACKAGENAME}-${OURPACKAGEVERSION}  in  ${OURBUILDDIR}  ..."
	echo '-----------------------------------------------------------------'
    fi

    time ( 

	mkdir "${OURBUILDDIR}/${OURPACKAGENAME}-${OURPACKAGEVERSION}"
	tar -cpf - . | (cd "${OURBUILDDIR}/${OURPACKAGENAME}-${OURPACKAGEVERSION}" && tar -xpvf -)

	(   cd "${OURBUILDDIR}/${OURPACKAGENAME}-${OURPACKAGEVERSION}" || exit 1
	    autoall
	    if [ "$DEBHELPERVERSION" -le 10 ]
	    then
	        cp -a debian-proto-9 debian
	    else
		cp -a debian-proto debian
	    fi

	    sed -i 's/RELEASE-TARGET/'"${RELEASE}"'/g' debian/changelog

	    DEBEMAIL="sysadm@rezopole.net" DEBFULLNAME="Jean-Daniel Pauget" dh_make -y --single --createorig
	    pandoc -f markdown -t plain --wrap=none README.md -o debian/README
	    cd debian && debuild -us -uc
	)
    )
    cp "${OURBUILDDIR}"/*deb "${FINALDEST}"
    echo '-----------------------------------------------------------------'
    ls *deb | while read FNAME
    do
	if echo "${FNAME}" | grep -- "-${RELEASE}\.deb$" 2>&1 > /dev/null
	then
	    ls -l "${FNAME}"
	else
	    FTARGET=$( echo "${FNAME}" | sed 's/\.deb$/'"-${RELEASE}.deb"'/' )
	    mv "${FNAME}" "${FTARGET}"
	    ls -l "${FTARGET}"
	fi
    done



asdig.cpp

// {
//
//  asdig, retreives Autonomous System (AS) infos from repositories via DNS rev-queries 
//  Copyright (C) 2016 Jean-Daniel Pauget <jdpauget@rezopole.net>
//  
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//  
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//  
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// }

#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/nameser.h>
#include <resolv.h>
#include <netdb.h>  // h_errno ???

#include <errno.h>  // errno ...
#include <string.h> //	strerror


#include <iostream>
#include <sstream>
#include <map>
#include <iomanip>

#include "level3addr.h"
#include "config.h"

using namespace std;

// the multipmap is used for sorting by prefix size

int resp_query (const char * query, multimap<int,string> &lout, const char * nameserver=NULL) {

    unsigned char answer [NS_PACKETSZ+10];


    if ((_res.options & RES_INIT) == 0) res_init();
    // _res.options |= RES_USEVC;   someday would use tcp rather than udp ?

    
    if(nameserver != NULL) {
	_res.nscount = 1;
	_res.nsaddr_list[0].sin_family = AF_INET;
	_res.nsaddr_list[0].sin_port = htons (53);
	inet_aton (nameserver, &(_res.nsaddr_list[0].sin_addr));
    }


    int r = res_query (query, ns_c_in, ns_t_txt, (unsigned char *)answer, NS_PACKETSZ);
    // int r = res_query (query, ns_c_in, ns_t_txt, (unsigned char *)answer, 30000);

    if (r < 0) {
	switch (h_errno) {
	    case HOST_NOT_FOUND:   /* Authoritative Answer Host not found */
		lout.insert (pair<int,string>(0, "HOST_NOT_FOUND"));
		return 0;
	    case TRY_AGAIN:        /* Non-Authoritative Host not found, or SERVERFAIL */
		lout.insert (pair<int,string>(0, "DNS_ERROR"));
		cerr << "res_query: " << query << " TRY_AGAIN" << endl;
		return 1;
	    case NO_RECOVERY:      /* Non recoverable errors, FORMERR, REFUSED, NOTIMP */
		lout.insert (pair<int,string>(0, "DNS_ERROR"));
		cerr << "res_query: " << query << "NO_RECOVERY:" << endl;
		return 1;
	    case NO_DATA:          /* Valid name, no data record of requested type */
		lout.insert (pair<int,string>(0, "NO_DATA"));
		return 0;
	    default:
		lout.insert (pair<int,string>(0, "DNS_ERROR"));
		cerr << "res_query: " << query << " error no:" << h_errno << endl;
		return 1;

	}
	int e = h_errno;
	cerr << "res_query error : " << e << " " << strerror(e) << endl;
	return -1;
    }

    ns_msg handle;  /* handle for response message */
    if (ns_initparse(answer, r, &handle) < 0) {
	int e = errno;
	cerr << "ns_initparse " << strerror(e) << endl;
        return -1;
    }

    ns_rr rr;   /* expanded resource record */
    for (int rrnum=0 ; rrnum < ns_msg_count(handle, ns_s_an) ; rrnum++) {
        if (ns_parserr(&handle, ns_s_an, rrnum, &rr) != 0) {
	    int e = errno;
	    cerr << "ns_parserr error : " << strerror (e) << endl;
        }
        if (ns_rr_type(rr) == ns_t_txt) {

	    int l = (int)(*ns_rr_rdata(rr)); // the length of the TXT field (? JD)
	    const unsigned char *pmax = ns_msg_base(handle) + ns_msg_size(handle);
	    const unsigned char *p = ns_rr_rdata(rr) + 1;
	    string s;
	    for (int i=0 ; (i<l) && (*p!=0) && (p<pmax) ; i++, p++) {
		s += (char)(*p);
	    }
// typical answer is :
// 199422 | 77.95.64.0/23 | FR | ripe | 2013-01-24 | REZOPOLE
	    int prefsize = 0;
	    {	size_t p,q;
		p = s.find('|');
		if (p != string::npos) {
		    p = s.find('/', p+1);
		    if (p != string::npos) {
			p++;
			q = s.find('|', p);
			if (q != string::npos) {
			    prefsize = 129 - atoi(s.substr(p,q-p).c_str());
			} else {
			    prefsize = 129 - atoi(s.substr(p).c_str());
			}
		    }
		}
	    }

	    lout.insert (pair<int,string>(prefsize, s));
	}
    }

    return 0;
}

using namespace rzpnet;

int main (int nb, char ** cmde) {

    int i;
    bool therewassomeerror = false;
    bool debug_query = false;
    const char 
	*rzpas = "asn.rezopole.net",
	*rzpv4 = "origin.asn.rezopole.net.",
	*rzpv6 = "origin6.asn.rezopole.net.",
	*cymruas = NULL,
	*cymruv4 = "origin.asn.cymru.com.",
	*cymruv6 = "origin6.asn.cymru.com.",
	*suffixas = rzpas,
	*suffixv4 = rzpv4,
	*suffixv6 = rzpv6,
	*nameserver = NULL;

    // first we look for nameserver
    bool dnsserver_multiple_change = false;
    for (i=1 ; i<nb ; i++) {
	if (cmde[i][0] == '@') {
	    if (nameserver != NULL) dnsserver_multiple_change = true;
	    nameserver = cmde[i] + 1;
	    continue;
	}
    }
    if (dnsserver_multiple_change)
	cerr << "WARNING: multiple dns servers given, will only use " << nameserver << " for dns-lookups" << endl;

    for (i=1 ; i<nb ; i++) {
	if (cmde[i][0] == '-')  {
	    if (strncmp (cmde[i], "-suffixv4=", 10) == 0) {
		suffixv4 = cmde[i] + 10;
		continue;
	    } else if (strncmp (cmde[i], "-suffixv6=", 10) == 0) {
		suffixv6 = cmde[i] + 10;
		continue;
	    } else if (strcmp (cmde[i], "-rezopole") == 0) {
		suffixas = rzpas;
		suffixv4 = rzpv4;
		suffixv6 = rzpv6;
		continue;
	    } else if (strcmp (cmde[i], "-cymru") == 0) {
		suffixas = cymruas;
		suffixv4 = cymruv4;
		suffixv6 = cymruv6;
		continue;
	    } else if (	    (strcmp (cmde[i], "-help") == 0)
			||  (strcmp (cmde[i], "--help") == 0)
			||  (strcmp (cmde[i], "-h") == 0)
		) {
		cout << "asdig [-rezopole] [-cymru] [-suffixv4=my.ipv4.suffix.com] [@IP.my.ser.ver]" << endl
		     << "      [-suffixv6=my.ipv6.suffix.net] IP|AS#### [ ... IP|AS### ... ] [-d|--debug]" << endl;
		return 0;
	    } else if (	    (strcmp (cmde[i], "-version") == 0)
			||  (strcmp (cmde[i], "--version") == 0)
			||  (strcmp (cmde[i], "-v") == 0)
		){
		cout << "asdig version " << ASDIG_VERSION << endl;
		return 0;
	    } else if (	(strcmp (cmde[i], "-d") == 0)
			|| (strcmp (cmde[i], "--debug") == 0)
		) {
		debug_query = true;
		continue;
	    } else {
		cerr << "unknown option \"" << cmde[i] << "\"" << endl;
		continue;
	    }
	} else if (cmde[i][0] == '@') {
	    continue;
	}

	stringstream q;
	Level3Addr adr;
	if ((strncmp ("as", cmde[i], 2) == 0) || (strncmp ("AS", cmde[i], 2) == 0)) {
	    q << "as" << (cmde[i]+2);
	    if (suffixas == NULL) {
		cerr << "falling back on rezopole for as-lookup" << endl;
		suffixas = rzpas;
	    }
	    q << '.' << suffixas;
	} else {
	    if (strchr(cmde[i], ':') != NULL) { // we have an IPv6
		adr = Level3Addr (TETHER_IPV6, string(cmde[i]));
	    } else if (strchr(cmde[i], '.') != NULL) { // we now assume IPv4
		adr = Level3Addr (TETHER_IPV4, string(cmde[i]));
	    }

	    adr.rev_arpa_radix (q);
	    switch (adr.t) {
		case TETHER_IPV4:
		    q << '.' << suffixv4;
		    break;
		case TETHER_IPV6:
		    q << '.' << suffixv6;
		    break;
		default:
		    break;
	    }
	}

	if (debug_query) {
	    cerr << "dig TXT " << q.str().c_str();
	    if (nameserver != NULL)
		cerr << " @" << nameserver;
	    cerr << endl;
	}

	multimap<int,string> l;
	if (  resp_query (q.str().c_str(), l, nameserver)  )
	    therewassomeerror = true;
	multimap<int, string>::iterator li;
	for (li=l.begin() ; li!=l.end() ; li++) {
	    cout << li->second << endl;
	}
    }

    if (therewassomeerror)
	return 1;

    return 0;
}

